src/script_opt/ScriptOpt.cc

```diff
@@ -657,4 +657,123 @@ void profile_script_execution() {
 
 void finish_script_execution() { profile_script_execution(); }
 
+// For now, we have equivalent concerns between ZAM and compile-to-C++.
+bool has_AST_node_unknown_to_script_opt(const ProfileFunc* prof, bool /* is_ZAM */) {
+    // Note that the following sets are not comprehensive across the
+    // standard tags, because some tags are only generated *by* script
+    // optimization
+    // clang-format off
+    static const std::set<StmtTag> known_stmts = {
+        STMT_PRINT,
+        STMT_EVENT,
+        STMT_EXPR,
+        STMT_IF,
+        STMT_WHEN,
+        STMT_SWITCH,
+        STMT_FOR,
+        STMT_NEXT,
+        STMT_BREAK,
+        STMT_RETURN,
+        STMT_LIST,
+        // STMT_EVENT_BODY_LIST,
+        STMT_INIT,
+        STMT_FALLTHROUGH,
+        STMT_WHILE,
+        // STMT_CATCH_RETURN,
+        // STMT_CHECK_ANY_LEN,
+        // STMT_CPP,
+        // STMT_ZAM,
+        STMT_NULL,
+        STMT_ASSERT,
+        // STMT_EXTERN,
+        // STMT_STD_FUNCTION,
+    };
+    // clang-format on
```

Comment:
I'm not so sure about the maintainability here.

Isn't it just `STMT_EXTERN` and `STMT_STD_FUNCTION` (and maybe STMT_CPP?) that need to be recognized from the original AST?

If you want to keep the full list, maybe some static assert / constexpr comparing NUM_STMTS and NUM_EXPRS with the number of statements listed here could detect inconsistencies at compilation?

Reply:
> Isn't it just `STMT_EXTERN` and `STMT_STD_FUNCTION` (and maybe STMT_CPP?) that need to be recognized from the original AST?

In principle it's more than that, per the others that are commented out. One could however rely on the fact that things like `STMT_CATCH_RETURN` shouldn't show up in a candidate for script optimization since it's only generated during script optimization, and then go with the simple list of what's-not-compilable. It seems though that that path is going to fail the next time one of those is added and forgets to update them here.

We could add comments to `StmtEnums.h` that when adding a node there, it should be added to `ScriptOpt.cc` too. Do you prefer that?

Reply:
I've added a commit using `static_assert()` so that compilation fails if new Stmt or Expr enums are added. I'll merge this PR if CI looks sensible.